engine/engine.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package engine
+
+import (
+	"os"
+
+	"github.com/rs/zerolog"
+)
+
+// Engine is an engine that is composed of multiple components.
+type Engine struct {
+	log        zerolog.Logger
+	components []*component
+
+	interrupt chan os.Signal
+	notify    chan error
+}
+
+// New creates a new engine.
+func New(log zerolog.Logger, name string, interrupt chan os.Signal) *Engine {
+	e := Engine{
+		log:       log.With().Str("engine", name).Logger(),
+		interrupt: interrupt,
+	}
+
+	return &e
+}
+
+// Component registers a new component for the engine. Components will be shut down
+// in the same order as the one in which they were registered.
+func (e *Engine) Component(name string, run func() error, stop func()) *Engine {
+	c := component{
+		log:  e.log.With().Str("component", name).Logger(),
+		run:  run,
+		stop: stop,
+	}
+
+	e.components = append(e.components, &c)
+
+	return e
+}
+
+// Run launches the engine components and waits for them to either finish successfully,
+// fail, or for an external signal to shut the engine down.
+func (e *Engine) Run() error {
+	e.notify = make(chan error, len(e.components))
+	for _, component := range e.components {
+		go component.Run(e.notify)
+	}
+
+	// Here, we are waiting for a signal, or for one of the components to fail
+	// or finish. In both cases, we proceed to shut down everything, while also
+	// entering a goroutine that allows us to force shut down by sending
+	// another signal.
+	select {
+	case <-e.interrupt:
+		e.log.Info().Msg("engine stopping")
+		e.stop()
+	case err := <-e.notify:
+		if err != nil {
+			e.log.Error().Err(err).Msg("engine failed")
+			e.stop()
+			return err
+		}
+		e.log.Info().Msg("engine done")
+	}
+	go func() {
```

Comment:
I don't think this would work as intended, because select is a blocking operation. I think this goroutine should be executed inside the `select`. The reason for this is that `e.stop()` can take some time to finish execution (in case of some long-running tasks or smthg), so forcing the program to stop should not depend on background processes.

```
func (e *Engine) Run() error {
    // ...

    select {
	case <-e.interrupt:
                go forceQuit()
		e.log.Info().Msg("engine stopping")
		e.stop()
	case err := <-e.notify:
                 go forceQuit()
		if err != nil {
			e.log.Error().Err(err).Msg("engine failed")
			e.stop()
			return err
		}
		e.log.Info().Msg("engine done")
	}
     return nil
}	
 
func forceQuit() {
	<-e.interrupt
	e.log.Warn().Msg("forcing exit")
	os.Exit(1)
}()

Reply:
Nice catch! Indeed, the way it currently was was wrong :) I changed it so that `e.stop()` starts by launching a goroutine that waits for an interrupt however. Please LMK if that looks good to you!